src/extensions/unit/unitext_hooks.cpp

```diff
@@ -680,6 +681,285 @@ DECLARE_PATCH(_UnitClass_Per_Cell_Process_AutoHarvest_Assign_Harvest_Mission_Pat
+         */
+
+        if (unittypeext->IsTransformRequiresFullCharge && this_ptr->CurrentCharge < this_ptr->Class->MaxCharge) {
+
+            /**
+             *  We don't have enough charge, return false
+             */
+            return 0x006511A0;
+        }
+
+        this_ptr->Mark(MARK_UP);
+        this_ptr->Locomotor_Ptr()->Mark_All_Occupation_Bits(MARK_UP);
+
+        UnitClass* newunit = Create_Transform_Unit(this_ptr);
+
+        if (newunit != nullptr) {
+
+            /**
+             *  Creating transformed unit succeeded, erase the original unit and force function to return true
+             */
+            return 0x0065114C;
+        } else {
+
+            /**
+             *  Creating transformed unit failed. Re-mark our occupation bits and return false.
+             */
+            return 0x00651168;
+        }
+    }
+
+    /**
+     *  Continue to deployability check.
+     */
+original_code:
+    return 0x00650BC2;
+}
+
+
+/**
+ *  #issue-715
+ *
+ *  Transforms a unit to another unit when a transformable unit deploys.
+ *
+ *  @author: Rampastring
+ */
+DECLARE_PATCH(_UnitClass_Try_To_Deploy_Transform_To_Vehicle_Patch) {
+    GET_REGISTER_STATIC(UnitClass*, this_ptr, esi);
+    static int address;
+    address = _UnitClass_Try_To_Deploy_Transform_To_Vehicle_Patch_Func(this_ptr);
+    JMP(address);
+}
+
+
+/**
+ *  #issue-715
+ *
+ *  Hack to display the the correct cursor for transformable units
+ *  upon ACTION_SELF.
+ *
+ *  @author: Rampastring
+ */
+DECLARE_PATCH(_UnitClass_What_Action_Self_Check_For_Vehicle_Transform_Patch) {
+    GET_REGISTER_STATIC(UnitClass*, this_ptr, esi);
+    static UnitTypeClass* unittype;
+    static UnitTypeClassExtension* unittypeext;
+
+    unittype = this_ptr->Class;
+    unittypeext = Extension::Fetch<UnitTypeClassExtension>(unittype);
+
+    /**
+     *  Stolen bytes/code.
+     */
+    if (unittype->DeploysInto != nullptr) {
+        goto original_code;
+    }
+
+    /**
+     *  Check if this unit is able to transform into another unit.
+     *  If not, we don't have anything else to do here.
+     */
+    if (unittypeext->TransformsInto == nullptr) {
+        goto no_deploy_action;
+    }
+
+    /**
+     *  If this unit is able to transform to a different unit, check if it requires charge for it.
+     *  If it does, then check whether we have enough charge.
+     */
+    if (unittypeext->IsTransformRequiresFullCharge && this_ptr->CurrentCharge < this_ptr->Class->MaxCharge) {
+
+        /**
+         *  We don't have enough charge!
+         */
+        goto return_no_deploy;
+
+    } else if (this_ptr->entry_2A4()) {
+
+        /**
+         *  The unit is dying or under an EMP effect, don't allow it to transform.
+         */
+        goto return_no_deploy;
+    }
+
+
+    /**
+     *  We can transform and either don't need charge or we have enough of it, return ACTION_SELF
+     */
+return_self:
```

Comment:
A little bit of code duplication here. We could clean this up by using making a single return label, with the return value being a variable that can be set in the _jumping_ case.